src/slack/app.rs

```diff
@@ -0,0 +1,214 @@
+use crate::settings::Slack;
+use crate::settings::WhoAmI;
+use crate::update::update_slack;
+use crate::userid::UserId;
+use actix_cors::Cors;
+use actix_session::CookieSession;
+use actix_session::Session;
+use actix_web::client::Client;
+use actix_web::cookie::SameSite;
+use actix_web::dev::HttpServiceFactory;
+use actix_web::http;
+use actix_web::web;
+use actix_web::Error;
+use actix_web::HttpResponse;
+use actix_web::Responder;
+use cis_client::getby::GetBy;
+use cis_client::AsyncCisClientTrait;
+use cis_profile::schema::Profile;
+use futures::future;
+use futures::Future;
+use futures::IntoFuture;
+use log::info;
+use oauth2::basic::BasicClient;
+use oauth2::prelude::*;
+use oauth2::AuthUrl;
+use oauth2::ClientId;
+use oauth2::ClientSecret;
+use oauth2::CsrfToken;
+use oauth2::RedirectUrl;
+use oauth2::Scope;
+use oauth2::TokenUrl;
+use std::sync::Arc;
+use std::sync::RwLock;
+use ttl_cache::TtlCache;
+use url::Url;
+
+const AUTH_URL: &str = "https://slack.com/oauth/authorize";
+const TOKEN_URL: &str = "https://slack.com/api/oauth.access";
+
+#[derive(Deserialize)]
+pub struct Auth {
+    code: String,
+    state: String,
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+pub struct SlackUser {
+    name: String,
+    id: String,
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+pub struct SlackTeam {
+    id: String,
+}
+
+#[derive(Deserialize, Serialize, Debug)]
+pub struct SlackUriData {
+    slack_auth_params: String,
+    direct_message_uri: String,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct SlackTokenResponse {
+    ok: bool,
+    access_token: String,
+    scope: String,
+    user: SlackUser,
+    team: SlackTeam,
+}
+
+fn redirect(client: web::Data<Arc<BasicClient>>, session: Session) -> impl Responder {
+    let (authorize_url, csrf_state) = client.authorize_url(CsrfToken::new_random);
+    info!("Setting csrf_state: {}", csrf_state.secret().clone());
+    session
+        .set("csrf_state", csrf_state.secret().clone())
+        .map(|_| {
+            HttpResponse::Found()
+                .header(http::header::LOCATION, authorize_url.to_string())
+                .finish()
+        })
+}
+
+fn auth<T: AsyncCisClientTrait + 'static>(
+    cis_client: web::Data<T>,
+    user_id: UserId,
+    query: web::Query<Auth>,
+    slack_uri_data: web::Data<SlackUriData>,
+    session: Session,
+) -> Box<dyn Future<Item = HttpResponse, Error = Error>> {
+    let code = query.code.clone();
```

Comment:
not needed we can use `query.code` a few lines later directly without copying here.